### PR TITLE
[Blog] Introduction to Metaprogramming

### PR DESCRIPTION
**Pull request by pikender**
Review and Feedback Awaited

### CODE REVIEW

**Comment by pikender** [comment]
Need to add a _Disclaimer_ asking to skip, if already know meta-programming and experienced with writing & using Macros

File: _posts/2016-04-06-intro-to-macros.md [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+>
+Why another tutorial on already well-documented metaprogramming topic ?
+>
+**To revise and refresh only what we would refer again and again when reviewing Model, Router, View extension DSLs**
+
+Extension DSLs will be using below constructs to get the job done :)
+
+- [Module.register_attribute/3](http://elixir-lang.org/docs/stable/elixir/Module.html#register_attribute/3)
+  - Registers an attribute. By registering an attribute, a developer is able to customize how Elixir will store and accumulate the attribute values.
```
Comment (nimish-mehta) [inline _posts/2016-04-06-intro-to-macros.md]: can we link to attribute here ?

File: _posts/2016-04-06-intro-to-macros.md [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+>
+Provides an Extension Framework to support features not included in core as extensions.
+>
+Strives for un-obstrusive parallel development of NectarCommerce and Extensions
+
+NectarCommerce is committed to providing a ready-to-use e-commerce solution but the definition of 100% is different under different business domains. It aims to solve common use-cases as part of the project and relying on extension framework to tap the rest.
+
+# Metaprogramming in Elixir
+
+_Note:_ If you are already familiar and experienced with Elixir Metaprogramming, you can jump to [last section](#last-section) as it lists the Metaprogramming resources
```
Comment (preetibhalla) [inline _posts/2016-04-06-intro-to-macros.md]: 'As it lists the Metaprogramming resources' >>  remove 'the' from here

File: _posts/2016-04-06-intro-to-macros.md [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+and constructs used in upcoming blog when creating different extension DSL's.
```
Comment (preetibhalla) [inline _posts/2016-04-06-intro-to-macros.md]: is it just the next blog for which we are using 'upcoming'?

File: _posts/2016-04-06-intro-to-macros.md [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+**Elixir _docs_** are excellent resources to get familair with Metaprogramming:
```
Comment (preetibhalla) [inline _posts/2016-04-06-intro-to-macros.md]: familair?

File: _posts/2016-04-06-intro-to-macros.md [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+- [Quote and unquote](http://elixir-lang.org/getting-started/meta/quote-and-unquote.html)
+- [Macros](http://elixir-lang.org/getting-started/meta/macros.html)
+- [Domain Specific Languages](http://elixir-lang.org/getting-started/meta/domain-specific-languages.html)
+
+Please check **Understanding Elixir Macros _blog_ series** by [@sasajuric](https://twitter.com/sasajuric)
```
Comment (preetibhalla) [inline _posts/2016-04-06-intro-to-macros.md]: 'Check out' instead of 'Please check'

File: _posts/2016-04-06-intro-to-macros.md [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+- [Understanding Elixir Macros, Part 4 - Diving Deeper](http://theerlangelist.com/article/macros_4)
+- [Understanding Elixir Macros, Part 5 - Reshaping the AST](http://theerlangelist.com/article/macros_5)
+- [Understanding Elixir Macros, Part 6 - In-place Code Generation](http://theerlangelist.com/article/macros_6)
+
+[**_Book_** Metaprogramming Elixir](https://pragprog.com/book/cmelixir/metaprogramming-elixir) by [@chris_mccord](https://twitter.com/chris_mccord) is next step for deep dive into metaprogramming.
+
+>
+Why another tutorial on already well-documented metaprogramming topic ?
+>
+**To revise and refresh only what we would refer again and again when reviewing Model, Router, View extension DSLs**
```
Comment (preetibhalla) [inline _posts/2016-04-06-intro-to-macros.md]: we can use 'To revise and refresh something that we would refer time and again'

File: _posts/2016-04-06-intro-to-macros.md [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+
+Extension DSLs will be using below constructs to get the job done :)
+
+- [Module.register_attribute/3](http://elixir-lang.org/docs/stable/elixir/Module.html#register_attribute/3)
+  - Registers an attribute. By registering an attribute, a developer is able to customize how Elixir will store and accumulate the attribute values.
+- [Module.put_attribute/3](http://elixir-lang.org/docs/stable/elixir/Module.html#put_attribute/3)
+  - Puts an Erlang attribute to the given module with the given key and value
```
Comment (preetibhalla) [inline _posts/2016-04-06-intro-to-macros.md:72]: fullstop missing

File: _posts/2016-04-06-intro-to-macros.md [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+- [@before_compile](http://elixir-lang.org/docs/stable/elixir/Module.html)
+  - A hook `__before_compile__/1` that will be invoked before the module is compiled
+  - allows us to inject code into the module when its definition is complete
+- [\_\_using\_\_ hook](http://elixir-lang.org/docs/stable/elixir/Kernel.html#use/2)
+  - check the examples for the usage, context and best practises
```
Comment (preetibhalla) [inline _posts/2016-04-06-intro-to-macros.md]: practises?

Also, check the examples for usage...

File: _posts/2016-04-06-intro-to-macros.md (hunk 3 above)
Comment (pikender) [inline _posts/2016-04-06-intro-to-macros.md]: No.

Should I change to blog_s_ ?

File: _posts/2016-04-06-intro-to-macros.md [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+Check out **Understanding Elixir Macros _blog_ series** by [@sasajuric](https://twitter.com/sasajuric)
+
+- [Understanding Elixir Macros, Part 1 - Basics](http://theerlangelist.com/article/macros_1)
+- [Understanding Elixir Macros, Part 2 - Micro Theory](http://theerlangelist.com/article/macros_2)
+- [Understanding Elixir Macros, Part 3 - Getting into the AST](http://theerlangelist.com/article/macros_3)
+- [Understanding Elixir Macros, Part 4 - Diving Deeper](http://theerlangelist.com/article/macros_4)
+- [Understanding Elixir Macros, Part 5 - Reshaping the AST](http://theerlangelist.com/article/macros_5)
+- [Understanding Elixir Macros, Part 6 - In-place Code Generation](http://theerlangelist.com/article/macros_6)
+
+[**_Book_** Metaprogramming Elixir](https://pragprog.com/book/cmelixir/metaprogramming-elixir) by [@chris_mccord](https://twitter.com/chris_mccord) is next step for deep dive into metaprogramming.
```
Comment (preetibhalla) [inline _posts/2016-04-06-intro-to-macros.md]: is the next step for diving deeper into...

File: _posts/2016-04-06-intro-to-macros.md [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+
+>
+Why another tutorial on already well-documented metaprogramming topic ?
```
Comment (preetibhalla) [inline _posts/2016-04-06-intro-to-macros.md]: ..on an already..

File: _posts/2016-04-06-intro-to-macros.md [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+>
+**To revise and refresh something that we would refer time and again when reviewing Model, Router, View extension DSLs**
+
+Extension DSLs will be using below constructs to get the job done :)
+
+- [Module.register_attribute/3](http://elixir-lang.org/docs/stable/elixir/Module.html#register_attribute/3)
+  - Registers an attribute. By registering an [attribute](http://elixir-lang.org/getting-started/module-attributes.html), a developer is able to customize how Elixir will store and accumulate the attribute values.
+- [Module.put_attribute/3](http://elixir-lang.org/docs/stable/elixir/Module.html#put_attribute/3)
+  - Puts an Erlang attribute to the given module with the given key and value
```
Comment (preetibhalla) [inline _posts/2016-04-06-intro-to-macros.md:72]: ..into a given module with a given key-value pair.

(if we are talking about a specific module and key-value, then 'the' can be used)

File: _posts/2016-04-06-intro-to-macros.md [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+- [@before_compile](http://elixir-lang.org/docs/stable/elixir/Module.html)
+  - A hook `__before_compile__/1` that will be invoked before the module is compiled
+  - It allows us to inject code into the module when its definition is complete
+- [\_\_using\_\_ hook](http://elixir-lang.org/docs/stable/elixir/Kernel.html#use/2)
+  - Find the documentation under the `examples for the usage, context and best practices` in link above
```
Comment (preetibhalla) [inline _posts/2016-04-06-intro-to-macros.md]: Find the documentation under title `examples for the usage, context and best practices` in link above

if it's not a title : 

Find the documentation under `examples for the usage, context and best practices` in link above

File: _posts/2016-04-06-intro-to-macros.md [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+  - `use ModuleName` looks and invokes `__using__` macro defined in `ModuleName` module
+- [bind_quoted](http://elixir-lang.org/docs/stable/elixir/Kernel.SpecialForms.html#quote/2)
+  - Find the documentation under the `bind_quoted option` in the link above
```
Comment (preetibhalla) [inline _posts/2016-04-06-intro-to-macros.md]: Find the documentation under title `bind_quoted option` in the link above

if it's not a title :

 Find the documentation under `bind_quoted option` in the link above

File: _posts/2016-04-06-intro-to-macros.md [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+- [Code.ensure_loaded?/1](http://elixir-lang.org/docs/stable/elixir/Code.html#ensure_loaded?/1)
+  - Ensures the given module is loaded
+- [apply/3](http://elixir-lang.org/docs/stable/elixir/Kernel.html#apply/2)
+  - Invokes the given `function` from `module` with the array of arguments `args`.
+- [Macro.escape/1](http://elixir-lang.org/docs/stable/elixir/Macro.html#escape/2)
+  - Recursively escapes a value so it can be inserted into a syntax tree
+
+#### <a name="last-section">Metaprogramming pattern as used across extensions</a>
+
+There are three parts needed at minimum to create & use an extension effectively:
```
Comment (preetibhalla) [inline _posts/2016-04-06-intro-to-macros.md]: Minimum three parts are needed to create & use an extension effectively:

File: _posts/2016-04-06-intro-to-macros.md [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+
+- Library Code
+- Service Code
+- Consumer Code
+
+An extension and its use with Nectar can be viewed as Producer / Consumer relationship bound by a communication protocol.
+
+**Extension** which want to add a route, say a list of favorites, is a **Producer (Service Code)**.
```
Comment (preetibhalla) [inline _posts/2016-04-06-intro-to-macros.md]: which **wants** to add a route, say a list of favorites, is a Producer (Service Code)

**Pull request closed without merging**